pkg/debug/flareshipping/flareshipping.go

```diff
@@ -0,0 +1,179 @@
+package flareshipping
+
+import (
+	"bytes"
+	"context"
+	"crypto"
+	"crypto/ecdsa"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"os/user"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/kolide/krypto/pkg/echelper"
+	"github.com/kolide/launcher/ee/control"
+	"github.com/kolide/launcher/pkg/agent"
+	"github.com/kolide/launcher/pkg/agent/types"
+	"github.com/kolide/launcher/pkg/debug/checkups"
+	"github.com/kolide/launcher/pkg/launcher"
+)
+
+type flarer interface {
+	RunFlare(ctx context.Context, k types.Knapsack, flareStream io.Writer, runtimeEnvironment checkups.RuntimeEnvironmentType) error
+}
+
+func RunFlareShip(logger log.Logger, k types.Knapsack, flarer flarer, requestUrl string) error {
+	// make sure we have a url to upload to
+	uploadUrl, err := requestUploadUrl(k, requestUrl)
+	if err != nil {
+		return fmt.Errorf("requesting upload url: %w", err)
+	}
+
+	flareBuf := bytes.NewBuffer([]byte{})
+
+	// run flare
+	ctx := context.Background()
+	if err := flarer.RunFlare(ctx, k, flareBuf, checkups.InSituEnvironment); err != nil {
+		return fmt.Errorf("running flare: %w", err)
+	}
+
+	flareUploadRequest, err := http.NewRequest(http.MethodPost, uploadUrl, flareBuf)
+	if err != nil {
+		return fmt.Errorf("creating request: %w", err)
+	}
+
+	flareUploadResponse, err := http.DefaultClient.Do(flareUploadRequest)
+	if err != nil {
+		return fmt.Errorf("sending request: %w", err)
+	}
+	defer flareUploadResponse.Body.Close()
+
+	flareUploadResponseBody, err := io.ReadAll(flareUploadResponse.Body)
+	if err != nil {
+		return fmt.Errorf("reading response: %w", err)
+	}
+
+	if len(flareUploadResponseBody) > 0 {
+		logger.Log("flare upload response", string(flareUploadResponseBody))
+	}
+
+	if flareUploadResponse.StatusCode != http.StatusOK {
+		return fmt.Errorf("got %s status in response: %s", flareUploadResponse.Status, string(flareUploadResponseBody))
+	}
+
+	return nil
+}
+
+func requestUploadUrl(k types.Knapsack, requestUrl string) (string, error) {
+	data, err := uploadUrlRequestData(k)
+	if err != nil {
+		return "", fmt.Errorf("creating request body: %w", err)
+	}
+
+	request, err := uploadUrlRequest(k, requestUrl, data)
+	if err != nil {
+		return "", fmt.Errorf("creating request: %w", err)
+	}
+
+	response, err := http.DefaultClient.Do(request)
+	if err != nil {
+		return "", fmt.Errorf("sending request: %w", err)
+	}
+	defer response.Body.Close()
+
+	body, err := io.ReadAll(response.Body)
+	if err != nil {
+		return "", fmt.Errorf("reading response: %w", err)
+	}
+
+	if response.StatusCode != http.StatusOK {
+		return "", fmt.Errorf("got %s status in response: %s", response.Status, string(body))
+	}
+
+	return string(body), nil
```

Comment:
I suspect we'll end up with some json response here -- we get more rope for future expansion.